File: app/src/main/java/com/example/speedtest_rework/core/download/Downloader.java
package com.example.speedtest_rework.core.download;


import com.example.speedtest_rework.core.base.Connection;

import java.io.InputStream;
import java.net.URL;

import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLSocketFactory;


public abstract class Downloader extends Thread {
    private Connection c;
    private String path;
    private int ckSize;
    private boolean stopASAP = false, resetASAP = false;
    private long totDownloaded = 0;

    public Downloader(Connection c, String path, int ckSize) {
        this.c = c;
        this.path = path;
        this.ckSize = Math.max(ckSize, 1);
        start();
    }

    private static final int BUFFER_SIZE = 10240;

    public void run() {
        try {

            long lastProgressEvent = System.currentTimeMillis();
            long ckBytes = ckSize * 1048576L, newRequestThreshold = ckBytes / 4;
            long bytesLeft = 0;
            URL url = new URL(path);
            HttpsURLConnection httpsConn = (HttpsURLConnection) url.openConnection();
            httpsConn.setSSLSocketFactory((SSLSocketFactory) SSLSocketFactory.getDefault());
            httpsConn.setRequestProperty("Cache-Control", "no-cache");
            httpsConn.setRequestProperty("Expires", "-1");
            httpsConn.setRequestProperty("Pragma", "no-cache");
            httpsConn.setRequestProperty("Connection", "Keep-Alive");
            httpsConn.setHostnameVerifier((hostname, session) -> true);
            httpsConn.connect();
            InputStream in = httpsConn.getInputStream();
            byte[] buf = new byte[BUFFER_SIZE];
            while (!stopASAP) {
                if (bytesLeft <= newRequestThreshold) {
                    bytesLeft += ckBytes;
                }
                int l = in.read(buf);
                if (l < 0) {
                    break;
                }
                bytesLeft -= l;
                if (resetASAP) {
                    totDownloaded = 0;
                    resetASAP = false;
                }
                totDownloaded += l;
                if (System.currentTimeMillis() - lastProgressEvent > 200) {
                    lastProgressEvent = System.currentTimeMillis();
                    onProgress(totDownloaded);
                }
            }
            c.close();
        } catch (Throwable t) {
            try {
                c.close();
            } catch (Throwable t1) {
            }
            onError(t.toString());
        }
    }

    public void stopASAP() {
        this.stopASAP = true;
    }

    public abstract void onProgress(long downloaded);

    public abstract void onError(String err);

    public void resetDownloadCounter() {
        resetASAP = true;
    }

    public long getDownloaded() {
        return resetASAP ? 0 : totDownloaded;
    }
}

File: app/src/main/java/com/example/speedtest_rework/core/download/Downloader.kt
import android.net.SSLCertificateSocketFactory
import kotlinx.coroutines.*
import java.io.InputStream
import java.net.URL
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLSocketFactory

abstract class Downloader(
    private val path: String, private val ckSize: Int,
) {
    private var totDownloaded: Long = 0

    private var job: Job? = null
    private var coroutineScope: CoroutineScope? = null
    private var inputStream: InputStream? = null


    fun startDownload() {
        stopASAP = true
        coroutineScope = CoroutineScope(Dispatchers.IO)
        job = coroutineScope?.launch {
            try {
                var lastProgressEvent = System.currentTimeMillis()
                val ckBytes = ckSize * 1024
                var bytesLeft: Long = 0
                val url = URL(path)
                val httpsConn = url.openConnection() as HttpsURLConnection
//                httpsConn.sslSocketFactory = SSLCertificateSocketFactory.getInsecure(0, null)
                httpsConn.sslSocketFactory = SSLSocketFactory.getDefault() as SSLSocketFactory
                httpsConn.setRequestProperty("Cache-Control", "no-cache")
                httpsConn.setRequestProperty("Expires", "-1")
                httpsConn.setRequestProperty("Pragma", "no-cache")
                httpsConn.setRequestProperty("Connection", "Keep-Alive")
                httpsConn.connect()
                inputStream = httpsConn.inputStream
                val buf = ByteArray(BUFFER_SIZE)
                while (true) {
                    bytesLeft += ckBytes
                    val l = inputStream?.read(buf)
                    if (l != null) {
                        if (l < 0 || !stopASAP) {
                            break
                        }
                        bytesLeft -= l.toLong()
                        totDownloaded += l.toLong()
                        if (System.currentTimeMillis() - lastProgressEvent > 200) {
                            lastProgressEvent = System.currentTimeMillis()
                            onProgress(totDownloaded)
                        }
                    }

                    if (!stopASAP) {
                        break
                    }


                }
                inputStream?.close()
            } catch (t: Throwable) {
                try {
                    inputStream?.close()
                    cancelDownload()
                } catch (t1: Throwable) {
                }
                onError(t.printStackTrace().toString())
            }
        }
    }

    fun cancelDownload() {
        coroutineScope?.cancel()
        job?.cancel()
        stopASAP = false
    }

    abstract fun onProgress(downloaded: Long)
    abstract fun onError(err: String?)
    fun resetDownloadCounter() {
        totDownloaded = 0
    }

    companion object {
        var stopASAP = true
        private const val BUFFER_SIZE = 1024
    }
}